cmd/kubeadm/app/phases/kubelet/kubelet.go

```diff
@@ -55,9 +56,25 @@ func CreateBaseKubeletConfiguration(cfg *kubeadmapi.MasterConfiguration, client
 		return err
 	}
 
+	var configmapName string
+	k8sVer, k8sErr := version.ParseSemantic(cfg.KubernetesVersion)
```

Comment:
The idea in this PR is:
**For master and `init`:**
If kubernetes version is configured and can be parsed, use this version as the post-fix of configmap name, e.g., "-1.9"; else use kubeadm version.
**For node and `join`:**
If kubelet version of the node can be gotten, use this version when getting configmap, else use kubeadm version.

But this will cause issues in below cases:
Suppose kubeadm version is `1.10.x` and kubernetes version is `1.9.x` and kubelet version is `1.10.x`:
- master parses kubernetes version successfully and writes configmap version as `1.9`;
- node gets kubelet version as `1.10` and tries to get `1.10` configmap, then fails.

Suppose kubeadm version is `1.10.x` and kubernetes version is `1.9.x` and kubelet version is `1.9.x`:
- master parses kubernetes version failed but gets kubeadm version successfully and writes configmap version as `1.10`;
- node gets kubelet version as `1.9` and tries to get `1.9` configmap, then fails.

Suppose kubeadm version is `1.10.x` and kubernetes version is `1.10.x` and kubelet version is `1.9.x`:
- master parses kubernetes version or gets kubeadm version successfully and writes configmap version as `1.10`;
- node gets kubelet version as `1.9` and tries to get `1.9` configmap, then fails.

@luxas @timothysc any suggestions? cc @kubernetes/sig-cluster-lifecycle-pr-reviews 

/hold

Reply:
I hope kubernetes version (master) at `1.9.x` with a kubelet at `1.10.x` is impossible in all deployment scenarios. Masters should always be upgraded before clients. 

I don't see why should the kubeadm version or the master version should ever be used to determine the config the kubelet gets. The kubelet version is what matters.

Reply:
>I hope kubernetes version (master) at 1.9.x with a kubelet at 1.10.x is impossible in all deployment scenarios. Masters should always be upgraded before clients.

agreed.

>I don't see why should the kubeadm version or the master version should ever be used to determine the config the kubelet gets. The kubelet version is what matters.

(base) kubelet config is embedded in `MasterConfiguration` of kubeadm, so I think the kubeadm version matters in this case. The kubelet version is actually what matters though.

Reply:
Are we just talking about the kubelet on the master here? I thought we were talking about all kubelets (masters and nodes).

Reply:
No, we are talking about all kubelets as you thought :)
Now I'm thinking of whether we need to enable the DKC feature in kubeadm. If so, what benifits does this bring us? Currently what kubeadm does wrt DKC is writing/uploading the base kubelet configuration and making the nodes consume it. But since `--init-config-dir` has been removed and `--config` is supported, seems like the base kubelet configuration is no longer needed. Maybe we want to remove the DKC support in kubeadm? @mtaufen @luxas WDYT?

Reply:
Shouldn't kubeadm install the config file for `--config` (which is the base kubelet configuration)?

Reply:
It should. (https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#kubelet-drop-in)

But I'm not sure: 
1. if kubeadm still should configure anything for kubelet in `MasterConfiguration.KubeletConfiguration` and create the base ConfigMap.
2. if kubeadm should create an base ConfigMap via the config file for `--config`; 

I think kubelet itself will create the ConfigMap from its `--config` file so that kubeadm don't need to do anything for DKC. (I don't follow the very details of the latest DKC/`--config` mechanism. Feel free to correct me if I'm wrong)

Reply:
The Kubelet does not create ConfigMaps nor does it configure Node.Spec.ConfigSource (the latter is blocked by the noderestriction admission controller - it's effectively a privilege escalation).

It may be better for Kubeadm to focus on `--config` and hold off trying to manage DKC for now - at least until DKC is beta. There is coordination involved with DKC that may be more appropriate to an always-on controller - like what is being designed in the Machines API.

Reply:
>It may be better for Kubeadm to focus on --config

@mtaufen do you mean migrating `/etc/systemd/system/kubelet.service.d/10-kubeadm.conf` to use `--config`, or anything else?

>and hold off trying to manage DKC for now - at least until DKC is beta. 

agreed.